README.md
# CurveDecomposition

Source of code associated to submitted RRPR paper.


To install the program see <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/INSTALL.txt">INSTALL.txt</a> file


If you want you can also directly test the programm online:

http://ipol-geometry.loria.fr/~phuc/ipol_demo/RRPR_demo/


* [![Build Status](https://travis-ci.org/ngophuc/CurveDecomposition.svg?branch=master)](https://travis-ci.org/ngophuc/CurveDecomposition)

# Examples

<p>File <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Samples/pentagonNoise4.sdp">pentagonNoise4.sdp</a>: </p>&#x000A;&#x000A;
<pre class="code highlight js-syntax-highlight plaintext">
<code>./testContourDecom -i ../Samples/pentagonNoise4.sdp -o ../Results/pentagonNoise4 -d ../ImaGene-forIPOL &#x000A; --maxScale 10 --samplingStep 1.0 --alphaMax 0.78 --thickness 0.2 --isseTol 4.0 --nbPointCircle 3</code>
</pre>&#x000A;&#x000A;
<p>
	<table cellpadding="5">
		<tr>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/pentagonNoise4.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/pentagonNoise4.png" alt="Input curve" />
			</a>	
		<br />
		Input curve
		</td>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/pentagonNoise4_OnlyArcSeg.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/pentagonNoise4_OnlyArcSeg.png" alt="Decomposition result" />
			</a>
		<br />
		Decomposition result
		</td>
		</tr>
	</table>
</p>

<p>File <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Samples/circle50.sdp">circle50.sdp</a>: </p>&#x000A;&#x000A;
<pre class="code highlight js-syntax-highlight plaintext">
<code>./testContourDecom -i ../Samples/circle50.sdp -o ../Results/circle50 -d ../ImaGene-forIPOL &#x000A; --maxScale 10 --samplingStep 1.0 --alphaMax 0.78 --thickness 0.2 --isseTol 4.0 --nbPointCircle 3</code>
</pre>&#x000A;&#x000A;
<p>
	<table cellpadding="5">
		<tr>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50.png" alt="Input curve" />
			</a>	
		<br />
		Input curve
		</td>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50_OnlyArcSeg.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50_OnlyArcSeg.png" alt="Decomposition result" />
			</a>
		<br />
		Decomposition result
		</td>
		</tr>
	</table>
</p>

<p>File <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Samples/circle50Noise4.sdp">circle50Noise4.sdp</a>: </p>&#x000A;&#x000A;
<pre class="code highlight js-syntax-highlight plaintext">
<code>./testContourDecom -i ../Samples/circle50Noise4.sdp -o ../Results/circle50Noise4 -d ../ImaGene-forIPOL &#x000A; --maxScale 10 --samplingStep 2.0 --alphaMax 0.78 --thickness 0.2 --isseTol 4.0 --nbPointCircle 3</code>
</pre>&#x000A;&#x000A;
<p>
	<table cellpadding="5">
		<tr>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50Noise4.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50Noise4.png" alt="Input curve" />
			</a>	
		<br />
		Input curve
		</td>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50Noise4_OnlyArcSeg.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/circle50Noise4_OnlyArcSeg.png" alt="Decomposition result" />
			</a>
		<br />
		Decomposition result
		</td>
		</tr>
	</table>
</p>

<p>File <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Samples/CircleRect.sdp">CircleRect.sdp</a>: </p>&#x000A;&#x000A;
<pre class="code highlight js-syntax-highlight plaintext">
<code>./testContourDecom -i ../Samples/CircleRect.sdp -o ../Results/CircleRect -d ../ImaGene-forIPOL &#x000A; --maxScale 10 --samplingStep 0.2 --alphaMax 0.78 --thickness 0.4 --isseTol 4.0 --nbPointCircle 3</code>
</pre>&#x000A;&#x000A;
<p>
	<table cellpadding="5">
		<tr>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/CircleRect.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/CircleRect.png" alt="Input curve" />
			</a>	
		<br />
		Input curve
		</td>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/CircleRect_OnlyArcSeg.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/CircleRect_OnlyArcSeg.png" alt="Decomposition result" />
			</a>
		<br />
		Decomposition result
		</td>
		</tr>
	</table>
</p>

<p>File <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Samples/exemple13_KNoise_03.sdp">exemple13_KNoise_03.sdp</a>: </p>&#x000A;&#x000A;
<pre class="code highlight js-syntax-highlight plaintext">
<code>./testContourDecom -i ../Samples/exemple13_KNoise_03.sdp -o ../Results/exemple13_KNoise_03 -d ../ImaGene-forIPOL &#x000A; --maxScale 10 --samplingStep 1.0 --alphaMax 0.78 --thickness 0.2 --isseTol 4.0 --nbPointCircle 3</code>
</pre>&#x000A;&#x000A;
<p>
	<table cellpadding="5">
		<tr>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/exemple13_KNoise_03.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/exemple13_KNoise_03.png" alt="Input curve" />
			</a>	
		<br />
		Input curve
		</td>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/exemple13_KNoise_03_OnlyArcSeg.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/exemple13_KNoise_03_OnlyArcSeg.png" alt="Decomposition result" />
			</a>
		<br />
		Decomposition result
		</td>
		</tr>
	</table>
</p>

<p>File <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Samples/symbol004_KNoise_05.sdp">symbol004_KNoise_05.sdp</a>: </p>&#x000A;&#x000A;
<pre class="code highlight js-syntax-highlight plaintext">
<code>./testContourDecom -i ../Samples/symbol004_KNoise_05.sdp -o ../Results/symbol004_KNoise_05 -d ../ImaGene-forIPOL &#x000A; --maxScale 10 --samplingStep 1.0 --alphaMax 0.78 --thickness 0.2 --isseTol 4.0 --nbPointCircle 3</code>
</pre>&#x000A;&#x000A;
<p>
	<table cellpadding="5">
		<tr>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/symbol004_KNoise_05.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/symbol004_KNoise_05.png" alt="Input curve" />
			</a>	
		<br />
		Input curve
		</td>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/symbol004_KNoise_05_OnlyArcSeg.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/symbol004_KNoise_05_OnlyArcSeg.png" alt="Decomposition result" />
			</a>
		<br />
		Decomposition result
		</td>
		</tr>
	</table>
</p>

<p>File <a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Samples/example7_KNoise_07.sdp">example7_KNoise_07.sdp</a>: </p>&#x000A;&#x000A;
<pre class="code highlight js-syntax-highlight plaintext">
<code>./testContourDecom -i ../Samples/example7_KNoise_07.sdp -o ../Results/example7_KNoise_07 -d ../ImaGene-forIPOL &#x000A; --maxScale 10 --samplingStep 1.0 --alphaMax 0.78 --thickness 0.2 --isseTol 4.0 --nbPointCircle 3</code>
</pre>&#x000A;&#x000A;
<p>
	<table cellpadding="5">
		<tr>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/example7_KNoise_07.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/example7_KNoise_07.png" alt="Input curve" />
			</a>	
		<br />
		Input curve
		</td>
		<td align="center" valign="center">
			<a href="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/example7_KNoise_07_OnlyArcSeg.pdf">
				<img width="150" src="https://github.com/ngophuc/CurveDecomposition/blob/master/Results/example7_KNoise_07_OnlyArcSeg.png" alt="Decomposition result" />
			</a>
		<br />
		Decomposition result
		</td>
		</tr>
	</table>
</p>

ImaGene/ChangeLog
2010-04-21 dcoeurjo
  * New methods in C4CSegment / C4CSegmentPencil to extract the segment/pencil Measure

2010-03-24 dcoeurjo
	* MeasureOfStraightLines: new Class added to compute the measure of sets of lines in the (a,b) parameter space

2010-03-23 00:18  kerautret
	* /tests/TestPAMI/displayFlatCurved.cxx:	  Rajout	  aff	  Curved	  info	  (pour	  A.V.)

2010-03-16 10:35  kerautret
	* /src/helper/DrawingXFIG.cxx:	  Rajout	  option	  dans	  drawingXFIG

2010-03-16 10:35  kerautret
	* /include/ImaGene/helper/DrawingXFIG.h:	  Rajout	  option	  dans	  drawingXFIG

2010-03-16 08:43  kerautret
	* /include/ImaGene/helper/DrawingXFIG.h:	  Rajout	  option	  dans	  drawingXFIG

2010-03-16 08:42  kerautret
	* /src/helper/DrawingXFIG.cxx:	  Rajout	  option	  dans	  drawingXFIG

2010-03-08 17:12  lachaud
	* /bin/freeman2polygon.cxx,	  /include/ImaGene/dgeometry2d/DMLPContour.h,	  /include/ImaGene/dgeometry2d/DMLPContour.ih,	  /include/ImaGene/dgeometry2d/FreemanChain.h,	  /include/ImaGene/dgeometry2d/FreemanChain.ih,	  /src/dgeometry2d/DMLPContour.cxx,	  /src/dgeometry2d/FreemanChain.cxx,	  /tests/test_DMLP.cxx:	  JOL:	  correct	  digital	  contour	  reconstruction	  from	  MLP

2010-03-04 15:50  kerautret
	* /include/ImaGene/helper/ShapeHelper.h:	  Modif	  des	  paramètres	  dans	  ShapeHelper:	  importFromPGM3d

2010-03-04 15:50  kerautret
	* /src/helper/ShapeHelper.cxx:	  Modif	  des	  paramètres	  dans	  ShapeHelper:	  importFromPGM3d

2010-03-01 16:43  lachaud
	* /include/ImaGene/base/Vector2i.h,	  /include/ImaGene/base/Vector2i.ih,	  /include/ImaGene/dgeometry2d/DMLPContour.h,	  /include/ImaGene/dgeometry2d/DMLPContour.ih,	  /tests/test_DMLP.cxx:	  JOL:	  modifs	  in	  contour	  reconstruction	  from	  MLP

2010-03-01 16:42  lachaud
	* /include/ImaGene/base/UndoableList.ih:	  JOL:	  correct	  bug	  in	  template	  specialization	  in	  namespaces

2010-02-19 13:20  lachaud
	* /include/ImaGene/base/OrderedAlphabet.h,	  /include/ImaGene/dgeometry2d/DMLPContour.h,	  /include/ImaGene/dgeometry2d/DMLPContour.ih,	  /include/ImaGene/dgeometry2d/FreemanChain.h,	  /include/ImaGene/mathutils/Mathutils.h,	  /src/base/OrderedAlphabet.cxx,	  /src/dgeometry2d/DMLPContour.cxx,	  /src/dgeometry2d/FreemanChain.cxx,	  /src/mathutils/Mathutils.cxx,	  /tests/test_DMLP.cxx:	  JOL:	  DMLP	  avec	  undo	  v3

2010-02-18 11:17  lachaud
	* /check/check_DMLP.sh,	  /include/ImaGene/base/UndoableList.h,	  /include/ImaGene/base/UndoableList.ih,	  /include/ImaGene/dgeometry2d/DMLPContour.h,	  /include/ImaGene/dgeometry2d/DMLPContour.ih,	  /src/CMakeLists.txt,	  /src/base/UndoableList.cxx,	  /src/dgeometry2d/DMLPContour.cxx,	  /tests/test_DMLP.cxx:	  JOL:	  DMLP	  undoable	  version

2010-02-14 17:47  kerautret
	* /tests/TestPAMI/CMakeLists.txt:	  Mise	  à	  jour	  modifs	  tests	  (TestPAMI	  et	  test_Profiles)

2010-02-14 17:43  kerautret
	* /tests/TestPAMI/ajusteImages.cxx,	  /tests/TestPAMI/compareCntDistance.cxx,	  /tests/TestPAMI/displayFlatCurved.cxx:	  Mise	  à	  jour	  modifs	  tests	  (TestPAMI	  et	  test_Profiles)

2010-02-14 17:41  kerautret
	* /src/helper/ShapeHelper.cxx:	  Mise	  à	  jour	  shape	  helper

2010-02-14 17:40  kerautret
	* /tests/TestPAMI/CMakeLists.txt,	  /tests/TestPAMI/compareShape.cxx,	  /tests/TestPAMI/displayBoxes.cxx,	  /tests/TestPAMI/displayNoise.cxx,	  /tests/test_Profiles.cxx:	  Mise	  à	  jour	  modifs	  tests	  (TestPAMI	  et	  test_Profiles)

2010-02-12 12:38  lachaud
	* /check,	  /check/check_DMLP.sh,	  /check/expected,	  /check/expected/chr1-dmlp-ccw.pts,	  /check/expected/chr1-dmlp-cw.pts,	  /check/expected/chr2-dmlp-ccw.pts,	  /check/expected/chr2-dmlp-cw.pts,	  /check/expected/chr3-dmlp-ccw.pts,	  /check/expected/chr3-dmlp-cw.pts,	  /check/expected/chr4-dmlp-ccw.pts,	  /check/expected/chr4-dmlp-cw.pts,	  /check/expected/circle-10-dmlp-ccw.pts,	  /check/expected/circle-10-dmlp-cw.pts,	  /check/expected/circle-dmlp-ccw.pts,	  /check/expected/circle-dmlp-cw.pts,	  /check/expected/circle-p-dmlp-ccw.pts,	  /check/expected/circle-p-dmlp-cw.pts,	  /check/expected/ellipseBruit-dmlp-ccw.pts,	  /check/expected/ellipseBruit-dmlp-cw.pts,	  /check/expected/ex1-dmlp-ccw.pts,	  /check/expected/ex1-dmlp-cw.pts,	  /check/expected/ex2-dmlp-ccw.pts,	  /check/expected/ex2-dmlp-cw.pts,	  /check/expected/ex3-dmlp-ccw.pts,	  /check/expected/ex3-dmlp-cw.pts,	  /check/expected/ex4-dmlp-ccw.pts,	  /check/expected/ex4-dmlp-cw.pts,	  /check/expected/ex5-dmlp-ccw.pts,	  /check/expected/ex5-dmlp-cw.pts,	  /check/expected/ex6-dmlp-ccw.pts,	  /check/expected/ex6-dmlp-cw.pts,	  /check/expected/ex7-dmlp-ccw.pts,	  /check/expected/ex7-dmlp-cw.pts,	  /check/expected/ex8-dmlp-ccw.pts,	  /check/expected/ex8-dmlp-cw.pts,	  /check/expected/flower-dmlp-ccw.pts,	  /check/expected/flower-dmlp-cw.pts,	  /check/expected/flower2-dmlp-ccw.pts,	  /check/expected/flower2-dmlp-cw.pts,	  /check/input,	  /check/input/chr1.fc,	  /check/input/chr2.fc,	  /check/input/chr3.fc,	  /check/input/chr4.fc,	  /check/input/circle-10.fc,	  /check/input/circle-p.fc,	  /check/input/circle.fc,	  /check/input/ellipseBruit.fc,	  /check/input/ex1.fc,	  /check/input/ex2.fc,	  /check/input/ex3.fc,	  /check/input/ex4.fc,	  /check/input/ex5.fc,	  /check/input/ex6.fc,	  /check/input/ex7.fc,	  /check/input/ex8.fc,	  /check/input/flower.fc,	  /check/input/flower2.fc,	  /check/output,	  /check/view_DMLP.sh,	  /include/ImaGene/dgeometry2d/DMLPContour.h,	  /include/ImaGene/dgeometry2d/DMLPContour.ih,	  /include/ImaGene/dgeometry2d/OctantChange.h,	  /include/ImaGene/dgeometry2d/OctantChange.ih,	  /include/ImaGene/mathutils/CFraction.ih,	  /src/CMakeLists.txt,	  /src/dgeometry2d/DMLPContour.cxx,	  /src/dgeometry2d/OctantChange.cxx,	  /tests/CTestTestfile.cmake,	  /tests/test_DMLP.cxx:	  JOL:	  DMLP	  1

2010-01-18 15:49  lachaud
	* /include/ImaGene/digitalnD/KnTypes.h:	  JOL:	  added	  parentheses	  in	  macros	  to	  avoid	  warnings	  in	  g++	  4.4.1

2010-01-08 17:19  lachaud

main.cpp
#include <iostream>

#include <boost/program_options/options_description.hpp>
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/variables_map.hpp>

#include "functions.h"
#include "decompositionAlgorithm.h"

namespace po=boost::program_options;
using namespace std;
const std::string version="1.0.1";
using namespace std;
using namespace DGtal;

int main(int argc, char *argv[])
{
    /* Read program parameters */
    double angleTol=10.0;
    po::options_description general_opt("Allowed options are: ");
    general_opt.add_options()
            ("help,h", "display this message")
            ("input,i", po::value<std::string>(), "input filename")
            ("imageneDir,d", po::value<std::string>(), "specify the ImaGene directory")
            ("sourceImageWidth", po::value<unsigned int>(), "source image width")
            ("sourceImageHeight", po::value<unsigned int>(), "source image height")
            ("output,o", po::value<std::string>()->default_value("./"), "output filename")
            ("version,v", "display the version number")
            ("eps,e","set output with eps format (default svg)")
            ("maxScale",  po::value<double>()->default_value(15.0), "set the maximal thickness for meaningful thickness detection")
            ("samplingStep",  po::value<double>()->default_value(1.0), "set the threshold of admissible angle in the tangent space")
            ("alphaMax,a",  po::value<double>()->default_value(M_PI/4.0), "set the threshold of admissible angle in the tangent space")
            ("thickness,t",  po::value<double>()->default_value(0.2), "set the tolerance for collinear test in the tangent space")
            ("nbPointCircle,n",  po::value<int>()->default_value(3), "set the minimum number of points on a circle")
            ("isseTol,s",  po::value<double>()->default_value(4.0), "set the ISSE tolerance approximated between segments and arc");

    bool parseOK=true;
    po::variables_map vm;
    try{
        po::store(po::parse_command_line(argc, argv, general_opt), vm);
    }catch(const std::exception& ex){
        trace.info()<< "Error checking program options: "<< ex.what()<< std::endl;
        parseOK=false;
    }
    if(vm.count("version")){
        trace.info() << "version: " << version << std::endl;
        return 0;
    }
    po::notify(vm);
    if(vm.count("help")||argc<=1|| !parseOK || !vm.count("input") || !vm.count("output") || !vm.count("imageneDir"))
    {
        trace.info()<< "Contour decomposition." <<std::endl << "Options: "<<std::endl
                    << general_opt << "\n";
        return 0;
    }
    string ImaGeneDIR = vm["imageneDir"].as<std::string>();
    double maxMT = vm["maxScale"].as<double>();
    double stepMT = vm["samplingStep"].as<double>();
    string inputFile=vm["input"].as<std::string>();
    string outputFile=vm["output"].as<std::string>();
    double alphaMax=vm["alphaMax"].as<double>();
    double thicknessMP=vm["thickness"].as<double>();
    double iseTol=vm["isseTol"].as<double>();
    int nbPointCircle=vm["nbPointCircle"].as<int>()-1;
    bool displayImageCanvas=vm.count("sourceImageWidth") && vm.count("sourceImageHeight");
    bool eps = vm.count("eps");
    unsigned int widthCanvas, heightCanvas=0;
    if (displayImageCanvas){
        widthCanvas=vm["sourceImageWidth"].as<unsigned int>();
        heightCanvas=vm["sourceImageHeight"].as<unsigned int>();
    }
    /* Read program parameters */

    /* Read contour */
    std::vector<std::vector<DGtal::Z2i::RealPoint> > vecTmp=PointListReader<DGtal::Z2i::RealPoint>::getPolygonsFromFile(inputFile);
    std::vector<std::vector<DGtal::Z2i::RealPoint> > vecPts=getContours(vecTmp);

    string outVecPts=outputFile+(eps? "OutPts.eps": "OutPts.svg");
    Board2D aBoard;
    aBoard.setLineWidth(100.0);
    aBoard.setFillColor(DGtal::Color::None);
    aBoard.setPenColor(DGtal::Color::Blue);
    for(auto it:vecPts)
        for(auto it_bis:it)
            aBoard.drawCircle(it_bis[0],it_bis[1],1);
    if(eps)
        aBoard.saveEPS(outVecPts.c_str());
    else
        aBoard.saveSVG(outVecPts.c_str());
    /* Read contour */

    /* Compute ATC decomposition */
    //call MT prog
    std::stringstream instruction;
    vector<vector<double> > vecMT;
    for (size_t it_contour = 0;it_contour != vecPts.size();it_contour++)
    {
        std::string contourFile=outputFile+"_"+std::to_string(it_contour)+".sdp";
        writeFile(vecPts.at(it_contour),contourFile.c_str());
        std::string noiseLevelMTFile=outputFile+"MeanThickness_"+std::to_string(it_contour)+".txt";
        if(vecPts.at(it_contour).size()<5*maxMT) {
            vector<double> oneVecMT;
            for (size_t it = 0;it != vecPts.at(it_contour).size(); it++)
                oneVecMT.push_back(1);
            vecMT.push_back(oneVecMT);
        }
        else {
            if(maxMT==stepMT) {
                ofstream myfile (noiseLevelMTFile);
                if (myfile.is_open())
                {
                    myfile << "# Noise levels exported from displayNoiseBS (source ImaGene):"<<endl;
                    myfile << "# Format X Y noiseLevel"<<endl;
                    for(size_t i=0; i<vecPts.at(it_contour).size(); i++)
                        myfile <<vecPts.at(it_contour).at(i)[0]<<" "<<vecPts.at(it_contour).at(i)[0]<<" "<<maxMT<<endl;
                    myfile.close();
                }
                else cout << "Unable to open file " <<noiseLevelMTFile<<endl;
            }
            else {
                instruction << ImaGeneDIR << "/build/tests/TestCompNoiseDetect/displayNoiseBS -srcPolygon " << contourFile
                            << " 0 1 CLOSED -setSampling "<<maxMT<<" "<<stepMT
                            << " -exportNoiseLevel "<< noiseLevelMTFile.c_str();
                std::system(instruction.str().c_str());
            }
            vector<double> oneVecMT=readMeanindfulThicknessFile(noiseLevelMTFile.c_str());
            vecMT.push_back(oneVecMT);
        }
    }
    // build ATC
    std::string outATC=outputFile+(eps? "ATC.eps": "ATC.svg");
    vector<vector<AlphaThickSegmentComputer2DD> > tangentCorverSet=adaptiveTangentCoverDecomposition(vecPts,vecMT,outATC.c_str());
    /* Compute ATC decomposition */

    /* Detect dominant points + optimization */
    vector<vector<RealPoint> > DP;
    std::string outDP=outputFile+(eps? "DP.eps": "DP.svg");
    DP=dominantPointDetection(tangentCorverSet,vecPts,outDP.c_str(),widthCanvas,heightCanvas);
    //Find index of DP on the curve
    vector<vector<int> > indexDP;
    for(size_t it_contour=0; it_contour<vecPts.size(); it_contour++)
    {
        vector<int> index;
        for(vector<RealPoint>::const_iterator it=DP.at(it_contour).begin(); it != DP.at(it_contour).end(); it++)
            index.push_back(findElement(vecPts.at(it_contour),*it));
        indexDP.push_back(index);
    }
    vector<vector<RealPoint> > newDP;
    std::string outNewDP=outputFile+(eps? "newDP.eps": "newDP.svg");
    newDP=dominantPointSimplification(DP,indexDP,vecPts,outNewDP,nbPointCircle);
    vector<vector<int> > indexNewDP;
    for(size_t it_contour=0; it_contour<vecPts.size(); it_contour++)
    {
        vector<int> index;
        for(vector<RealPoint>::const_iterator it=newDP.at(it_contour).begin(); it != newDP.at(it_contour).end(); it++)
            index.push_back(findElement(vecPts.at(it_contour),*it));
        indexNewDP.push_back(index);
    }
    /* Detect dominant points + optimization */

    /* transform DP into the tangent space */
    vector<vector<RealPoint> > MP=tangentSpaceTransform(newDP,vecPts);
    /* transform DP into the tangent space */

    /* Process arc and seg decomposition */
    vector<vector<double> > DecSegmentsPh,DecArcsPh;
    std::string outDec=outputFile+(eps? "ArcSeg.eps": "ArcSeg.svg");
    std::string outOnlyDec=outputFile+(eps? "OnlyArcSeg.eps": "OnlyArcSeg.svg");
    vector<vector<int> > isolatedPoint=arcSegmentDecomposition(vecPts,indexNewDP,MP,alphaMax,thicknessMP,iseTol,angleTol,nbPointCircle,DecSegmentsPh,DecArcsPh);
    drawDecomposition(vecPts,newDP,isolatedPoint,DecSegmentsPh,DecArcsPh,outDec.c_str(),outOnlyDec.c_str(),widthCanvas,heightCanvas);
    /* Process arc and seg decomposition */

    return 0;
}
